### PR TITLE
Timeline fixes & improvements

### PR DESCRIPTION
**Pull request by jiramarc**
# Product Requirements Document
## Annotation Timeline — Bug Fixes & UX Improvements

**Document Status:** Completed  
**Version:** 1.0  
**Date:** May 19, 2026  

---

## 1. Overview

This document describes the requirements and acceptance criteria for a set of bug fixes and UX improvements targeting the annotation timeline component. These issues affected annotation management, timeline navigation, scroll behavior, and zoom controls across the video viewport and right sidebar.

---

## 2. Background & Problem Statement

Users reported multiple behavioral issues with the annotation timeline that degraded the editing experience. These ranged from missing confirmation dialogs on destructive actions to incorrect scroll and zoom behavior, introducing risk of data loss and creating confusing interactions when navigating between the video viewport and the timeline.

---

## 3. Scope

**In scope:** Annotation timeline, video viewport annotation overlay, right sidebar annotation panel, and timeline zoom/scroll controls.

**Out of scope:** Annotation data model changes, backend API changes, and new annotation feature development.

---

### 4.1 Delete Confirmation for Single Annotation

**Problem:** When a user deletes a single annotation, no confirmation dialog is shown. This is inconsistent with expected behavior for destructive actions and creates risk of accidental data loss.

**Requirement:** A confirmation popup must appear whenever a user initiates deletion of a single annotation, prompting the user to confirm or cancel the action before it is executed.

**Acceptance Criteria:**
- Clicking the delete action on a single annotation triggers a confirmation dialog.
- The dialog clearly states which annotation will be deleted.
- Confirming the dialog proceeds with deletion; cancelling dismisses the dialog without any changes.
- The popup is consistent in style with other confirmation dialogs in the application.

---

### 4.2 Auto-Scroll to Newly Created Annotation

**Problem:** After creating a new annotation, the timeline does not automatically scroll to bring the newly created annotation into view, requiring the user to manually locate it.

**Requirement:** Upon creating a new annotation, the timeline must automatically scroll to the position of the newly created annotation.

**Acceptance Criteria:**
- Immediately after creation, the timeline scrolls to make the new annotation visible.
- The scroll behavior is smooth and does not cause flickering.
- The timeline scale (zoom level) is not affected by this scroll action.

---

### 4.3 Auto-Scroll to Selected Annotation from Viewport or Sidebar

**Problem:** When a user selects an annotation by clicking on its overlay in the video viewport, or by clicking it in the right sidebar, the timeline does not scroll to reflect that selection.

**Requirement:** Selecting an annotation from either the video viewport overlay or the right sidebar must cause the timeline to scroll to the corresponding annotation.

**Acceptance Criteria:**
- Clicking an annotation overlay in the video viewport scrolls the timeline to the selected annotation.
- Clicking an annotation in the right sidebar scrolls the timeline to the selected annotation.
- The timeline scale (zoom level) is not altered during this scroll action.
- The selected annotation is visually highlighted after the scroll.

---

### 4.4 Fix Timeline Flickering on Scroll

**Problem:** The timeline flickers when the user scrolls down, creating visual instability that disrupts the editing workflow.

**Requirement:** Timeline scrolling must be smooth and stable with no flickering under normal usage conditions.

**Acceptance Criteria:**
- Scrolling the timeline in any direction produces no visual flickering.
- Performance remains stable when scrolling through timelines with a large number of annotations.

---

### 4.5 Annotation Overlay Click Must Not Change Timeline Scale

**Problem:** Clicking on an annotation overlay in the video viewport and navigating the timeline back to the current frame also changes the timeline zoom scale. Root cause identified: the system was using a Focus action that fits the entire annotation duration into view, inadvertently altering the scale.

**Requirement:** When clicking on an annotation overlay in the viewport causes the timeline to navigate to the current frame, the timeline zoom scale must remain unchanged.

**Acceptance Criteria:**
- Clicking an annotation overlay in the viewport navigates the timeline to the relevant frame without changing the zoom scale.
- The Focus action (or equivalent) is not used for this navigation; a scroll-only approach is used instead.
- The timeline zoom level before and after the click remains identical.

---

### 4.6 Delete First Annotation in a Split Group Must Not Delete the Entire Group

**Problem:** When an annotation has been split into a group and the user deletes the first annotation in that group, the entire group is deleted instead of only the individual annotation.

**Requirement:** Deleting any individual annotation within a split group must delete only that annotation and must not affect other annotations in the group.

**Acceptance Criteria:**
- Deleting the first annotation in a split group removes only that annotation.
- The remaining annotations in the group are unaffected and remain correctly displayed.
- Deletion of any other annotation in the group (second, third, etc.) also removes only that individual item.
- The confirmation dialog (see 4.1) is shown before deletion proceeds.

---

### 4.7 Fix Zoom In/Out Button Behavior on Timeline Slider

**Problem:** Clicking the zoom in or zoom out button on the timeline slider immediately jumps the timeline to its maximum zoom-out level, instead of incrementally adjusting the scale up or down.

**Requirement:** The zoom in and zoom out buttons must incrementally adjust the timeline scale by a defined step value per click, rather than jumping to an extreme value.

**Acceptance Criteria:**
- Each click of the zoom in button increases the timeline scale by one consistent step.
- Each click of the zoom out button decreases the timeline scale by one consistent step.
- The zoom level never jumps to the minimum or maximum in a single click unless it is already one step away from that boundary.
- The slider position updates in sync with the zoom level after each click.

---

## 5. User Stories

| # | As a… | I want to… | So that… |
|---|--------|------------|----------|
| 1 | Video editor | See a confirmation before deleting an annotation | I don't accidentally lose annotation work |
| 2 | Video editor | Have the timeline scroll to a newly created annotation | I can immediately see and edit it without searching |
| 3 | Video editor | Have the timeline scroll when I select an annotation from the viewport or sidebar | I always know where the annotation sits in the timeline |
| 4 | Video editor | Scroll the timeline without flickering | The interface feels stable and professional |
| 5 | Video editor | Click an annotation overlay without the timeline zoom changing | My preferred zoom level is preserved during navigation |
| 6 | Video editor | Delete only one annotation from a split group | I can manage individual segments without affecting the rest |
| 7 | Video editor | Use zoom buttons to incrementally adjust the timeline scale | I have precise control over how much detail I see |

---

## 6. Out of Scope / Non-Goals

- Changes to how annotations are stored or structured in the data model.
- Redesign of the confirmation dialog component.
- Addition of undo/redo functionality (tracked separately).
- Changes to the annotation splitting or merging workflow beyond the bug fix in 4.6.

---

## 7. Dependencies

- Timeline scroll utilities must support programmatic scrolling without side effects on zoom state.
- The delete action handler must be refactored to distinguish between single-annotation deletion and group deletion.
- The Focus action used for viewport click navigation must be decoupled from the scroll-to-annotation behavior.

---

## 8. Risks & Considerations

| Risk | Likelihood | Mitigation |
|------|------------|------------|
| Refactoring Focus action may introduce regressions in other viewport interactions | Medium | Thorough regression testing across all viewport interaction paths |
| Scroll-to-annotation logic may conflict with manual scrolling state | Low | Implement debounce and only trigger on explicit selection events |
| Group deletion fix may require changes to how split groups are identified | Medium | Add unit tests for group membership and deletion boundary cases |

---

## 9. Acceptance Sign-Off Checklist

- [x] 4.1 — Confirmation popup shown on single annotation delete
- [x] 4.2 — Timeline scrolls to newly created annotation
- [x] 4.3 — Timeline scrolls when annotation selected from viewport or sidebar
- [x] 4.4 — No flickering during timeline scroll
- [x] 4.5 — Timeline scale unchanged when clicking annotation overlay in viewport
- [x] 4.6 — Deleting first annotation in split group deletes only that annotation
- [x] 4.7 — Zoom in/out buttons increment/decrement scale one step at a time

### CODE REVIEW

File: plugins/idah-video/frontend/src/lib/commands/annotation/split.ts [hunk 1]
```diff
@@ -60,8 +55,12 @@ export function register(driver: IIdahDriverV2): void {
         at = props.at;
       } else {
         // Shortcut invocation — derive from current selection and viewport
-        const sel = selection.value;
-        if (sel && sel.type === "annotation") {
+        if (selection.isAnnotation()) {
+          const sel = selection.value as {
+            type: "annotation";
+            annotation: IAnnotationRecord<Record<string, unknown>, Record<string, unknown>>;
+          };
```
Comment (rcserrano06) [inline plugins/idah-video/frontend/src/lib/commands/annotation/split.ts]: Would it be better to use the type:
```
const sel = selection.value as IAnnotationSelection;
```

File: plugins/idah-video/frontend/src/lib/commands/timeline/scroll-to-annotation.ts [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+          },
+          combine(p: any) {
+            return p;
+          },
+        };
+      }
+
+      const { startRange, endRange } = viewport.timeline.range;
+      const rangeWidth = endRange - startRange;
+      const newStart = Math.max(0, shape.start - 10);
```
Comment (rcserrano06) [inline plugins/idah-video/frontend/src/lib/commands/timeline/scroll-to-annotation.ts]: I think we can use the same margin computation as the one we use in focus?
```
const margin = Math.max(10, Math.round((shape.end - shape.start) * 0.1));
const newStart = Math.max(0, shape.start - margin);
```

**Review by rcserrano06 — CHANGES_REQUESTED** [review]

**Review by rcserrano06 — APPROVED** [review]
